pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,363 @@
+		"The total number of active series per user and attribution.", variableLabels[:len(variableLabels)-1],
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	return tracker
+}
+
+func (t *Tracker) hasSameLabels(labels []string) bool {
+	return slices.Equal(t.labels, labels)
+}
+
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func (t *Tracker) cleanupTrackerAttribution(key string) {
+	t.observedMtx.Lock()
+	defer t.observedMtx.Unlock()
+	delete(t.observed, key)
+}
+
+func (t *Tracker) IncrementActiveSeries(lbs labels.Labels, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, now, 1, 0, 0, nil, true)
+}
+
+func (t *Tracker) DecrementActiveSeries(lbs labels.Labels) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, time.Time{}, -1, 0, 0, nil, false)
+}
+
+func (t *Tracker) Collect(out chan<- prometheus.Metric) {
+
+	if t.isOverflow.Load() {
+		out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, t.overflowCounter.activeSerie.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, t.overflowCounter.receivedSample.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, t.overflowCounter.totalDiscarded.Load(), t.overflowLabels...)
+		return
+	}
+
+	// Collect metrics for all observed keys
+	t.observedMtx.RLock()
+	defer t.observedMtx.RUnlock()
+	for key, o := range t.observed {
+		if key == "" {
+			continue
+		}
+		keys := strings.Split(key, string(sep))
+
+		keys = append(keys, t.userID)
+		if o.activeSerie.Load() > 0 {
+			out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, o.activeSerie.Load(), keys...)
+		}
+		if o.receivedSample.Load() > 0 {
+			out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, o.receivedSample.Load(), keys...)
+		}
+		o.discardedSampleMtx.Lock()
+		for reason, discarded := range o.discardedSample {
+			out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, discarded.Load(), append(keys, reason)...)
+		}
+		o.discardedSampleMtx.Unlock()
+	}
+}
+
+func (t *Tracker) IncrementDiscardedSamples(lbs []mimirpb.LabelAdapter, value float64, reason string, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now, 0, 0, value, &reason, true)
+}
+
+func (t *Tracker) IncrementReceivedSamples(req *mimirpb.WriteRequest, now time.Time) {
+	if t == nil {
+		return
+	}
+
+	dict := make(map[string]int)
+	for _, ts := range req.Timeseries {
+		lvs := t.extractLabelValuesFromLabelAdapater(ts.Labels)
+		dict[t.hashLabelValues(lvs)] += len(ts.TimeSeries.Samples) + len(ts.TimeSeries.Histograms)
+	}
```

Comment:
This is the hottest path on our application, we should optimize it as much as possible.
Why do we need to build a new data structure (which escapes to heap) with holds `[]mimirpb.LabelAdapter` slices that escape to heap, which create a string that escapes to heap just to put it into a `dict` that I don't think should be a map, even if we need _some_ data structure.

Can we just extract the labelValues byte slices, recycled from a pool, and process each one separately in the loop below?

Reply:
as paired, addressed in this commit a2ffe5a893c9a9eef26160c760cc6e57e50a3f1d